File: trunk/src/applications/likwid-powermeter.c
/*
 * =======================================================================================
 *
 *      Filename:  likwid-powermeter.c
 *
 *      Description:  An application to get information about power 
 *      consumption on architectures implementing the RAPL interface.
 *
 *      Version:   <VERSION>
 *      Released:  <DATE>
 *
 *      Author:  Jan Treibig (jt), jan.treibig@gmail.com
 *      Project:  likwid
 *
 *      Copyright (C) 2013 Jan Treibig 
 *
 *      This program is free software: you can redistribute it and/or modify it under
 *      the terms of the GNU General Public License as published by the Free Software
 *      Foundation, either version 3 of the License, or (at your option) any later
 *      version.
 *
 *      This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *      WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *      PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License along with
 *      this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * =======================================================================================
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sched.h>
#include <sys/types.h>
#include <unistd.h>
#include <ctype.h>

#include <types.h>
#include <strUtil.h>
#include <error.h>
#include <lock.h>
#include <timer.h>
#include <cpuid.h>
#include <numa.h>
#include <accessClient.h>
#include <msr.h>
#include <affinity.h>
#include <perfmon.h>
#include <power.h>
#include <thermal.h>

/* #####   MACROS  -  LOCAL TO THIS SOURCE FILE   ######################### */

#define HELP_MSG \
printf("\nlikwid-powermeter --  Version  %d.%d \n\n",VERSION,RELEASE); \
printf("A tool to print Power and Clocking information on Intel SandyBridge CPUS.\n"); \
printf("Options:\n"); \
printf("-h\t Help message\n"); \
printf("-v\t Version information\n"); \
printf("-M\t set how MSR registers are accessed: 0=direct, 1=msrd \n"); \
printf("-c\t specify socket to measure\n"); \
printf("-i\t print information from MSR_PKG_POWER_INFO register and Turbo Mode\n"); \
printf("-s <duration>\t set measure duration in sec. (default 2s) \n"); \
printf("-p\t print dynamic clocking and CPI values\n\n");   \
printf("Usage: likwid-powermeter -s 4 -c 1 \n");  \
printf("Alternative as wrapper: likwid-powermeter -c 1 ./a.out\n")

#define VERSION_MSG \
printf("likwid-powermeter  %d.%d \n\n",VERSION,RELEASE)


int main (int argc, char** argv)
{
    int socket_fd = -1;
    int optInfo = 0;
    int optClock = 0;
    int optStethoscope = 0;
    int socketId = 0;
    double runtime;
    int hasDRAM = 0;
    int c;
    bstring argString;
    bstring eventString = bfromcstr("CLOCK");
    int numThreads=0;
    int threads[MAX_NUM_THREADS];

    while ((c = getopt (argc, argv, "+c:hiM:ps:v")) != -1)
    {
        switch (c)
        {
            case 'c':
                if (! (argString = bSecureInput(10,optarg)))
                {
                    fprintf(stderr,"Failed to read argument string!\n");
                    exit(EXIT_FAILURE);
                }

                socketId = str2int((char*) argString->data);
                bdestroy(argString);

                break;

            case 'h':
                HELP_MSG;
                exit (EXIT_SUCCESS);
            case 'i':
                optInfo = 1;
                break;
            case 'M':  /* Set MSR Access mode */
                CHECK_OPTION_STRING;
                accessClient_setaccessmode(str2int((char*) argString->data));
                break;
            case 'p':
                optClock = 1;
                break;
            case 's':
                CHECK_OPTION_STRING;
                optStethoscope = str2int((char*) argString->data);

                break;
            case 'v':
                VERSION_MSG;
                exit (EXIT_SUCCESS);
            case '?':
                if (isprint (optopt))
                {
                    fprintf (stderr, "Unknown option `-%c'.\n", optopt);
                }
                else
                {
                    fprintf (stderr,
                            "Unknown option character `\\x%x'.\n",
                            optopt);
                }
                return EXIT_FAILURE;
            default:
                HELP_MSG;
                exit (EXIT_SUCCESS);
        }
    }

    if (!lock_check())
    {
        fprintf(stderr,"Access to performance counters is locked.\n");
        exit(EXIT_FAILURE);
    }

    if (cpuid_init() == EXIT_FAILURE)
    {
        ERROR_PLAIN_PRINT(Unsupported processor!);
    }
    numa_init(); /* consider NUMA node as power unit for the moment */
    accessClient_init(&socket_fd);
    msr_init(socket_fd);
    timer_init();

    /* check for supported processors */
    if ((cpuid_info.model == SANDYBRIDGE_EP) ||
            (cpuid_info.model == SANDYBRIDGE) ||
            (cpuid_info.model == IVYBRIDGE) ||
            (cpuid_info.model == IVYBRIDGE_EP) ||
            (cpuid_info.model == HASWELL) ||
            (cpuid_info.model == NEHALEM_BLOOMFIELD) ||
            (cpuid_info.model == NEHALEM_LYNNFIELD) ||
            (cpuid_info.model == NEHALEM_WESTMERE))
    {
        power_init(numa_info.nodes[0].processors[0]);
    }
    else
    {
        fprintf (stderr, "Query Turbo Mode only supported on Intel Nehalem/Westmere/SandyBridge/IvyBridge/Haswell processors!\n");
        exit(EXIT_FAILURE);
    }

    double clock = (double) timer_getCpuClock();

    printf(HLINE);
    printf("CPU name:\t%s \n",cpuid_info.name);
    printf("CPU clock:\t%3.2f GHz \n",  (float) clock * 1.E-09);
    printf(HLINE);

    if (optInfo)
    {
        if (power_info.turbo.numSteps != 0)
        {
            printf("Base clock:\t%.2f MHz \n",  power_info.baseFrequency );
            printf("Minimal clock:\t%.2f MHz \n",  power_info.minFrequency );
            printf("Turbo Boost Steps:\n");
            for (int i=0; i < power_info.turbo.numSteps; i++ )
            {
                printf("C%d %.2f MHz \n",i+1,  power_info.turbo.steps[i] );
            }
        }
        printf(HLINE);
    }

    if (cpuid_info.model == SANDYBRIDGE_EP)
    {
        hasDRAM = 1;
    }
    else if ((cpuid_info.model != SANDYBRIDGE) &&
            (cpuid_info.model != SANDYBRIDGE_EP)  &&
            (cpuid_info.model != IVYBRIDGE)  &&
            (cpuid_info.model != IVYBRIDGE_EP)  &&
            (cpuid_info.model != HASWELL))
    {
        fprintf (stderr, "RAPL not supported on this processor!\n");
        exit(EXIT_FAILURE);
    }

    if (optInfo)
    {
        printf("Thermal Spec Power: %g Watts \n", power_info.tdp );
        printf("Minimum  Power: %g Watts \n", power_info.minPower);
        printf("Maximum  Power: %g Watts \n", power_info.maxPower);
        printf("Maximum  Time Window: %g micro sec \n", power_info.maxTimeWindow);
        printf(HLINE);
        exit(EXIT_SUCCESS);
    }

    if (optClock)
    {
        affinity_init();
        argString = bformat("S%u:0-%u", socketId, cpuid_topology.numCoresPerSocket-1);
        numThreads = bstr_to_cpuset(threads, argString);
        bdestroy(argString);
        perfmon_init(numThreads, threads, stdout);
        perfmon_setupEventSet(eventString, NULL);
    }

    {
        PowerData pDataPkg;
        PowerData pDataDram;
        int cpuId = numa_info.nodes[socketId].processors[0];
        printf("Measure on socket %d \n", socketId );

        if (optStethoscope)
        {

            if (optClock)
            {
                perfmon_startCounters();
            }

            if (hasDRAM) power_start(&pDataDram, cpuId, DRAM);
            power_start(&pDataPkg, cpuId, PKG);
            sleep(optStethoscope);
            power_stop(&pDataPkg, cpuId, PKG);
            if (hasDRAM) power_stop(&pDataDram, cpuId, DRAM);

            if (optClock)
            {
                perfmon_stopCounters();
                perfmon_printCounterResults();
                perfmon_finalize();
            }
            runtime = (double) optStethoscope;
        }
        else
        {
            TimerData time;
            argv +=  optind;
            bstring exeString = bfromcstr(argv[0]);

            for (int i=1; i<(argc-optind); i++)
            {
                bconchar(exeString, ' ');
                bcatcstr(exeString, argv[i]);
            }
            printf("%s\n",bdata(exeString));


            if (optClock)
            {
                perfmon_startCounters();
            }

            if (hasDRAM) power_start(&pDataDram, cpuId, DRAM);
            power_start(&pDataPkg, cpuId, PKG);
            timer_start(&time);

            if (system(bdata(exeString)) == EOF)
            {
                fprintf(stderr, "Failed to execute %s!\n", bdata(exeString));
                exit(EXIT_FAILURE);
            }
            timer_stop(&time);
            power_stop(&pDataPkg, cpuId, PKG);
            if (hasDRAM) power_stop(&pDataDram, cpuId, DRAM);

            if (optClock)
            {
                perfmon_stopCounters();
                perfmon_printCounterResults();
                perfmon_finalize();
            }
            runtime = timer_print(&time);
        }

        printf("Runtime: %g s \n",runtime);
        printf("Domain: PKG \n");
        printf("Energy consumed: %g Joules \n", power_printEnergy(&pDataPkg));
        printf("Power consumed: %g Watts \n", power_printEnergy(&pDataPkg) / runtime );
        if (hasDRAM)
        {
            printf("Domain: DRAM \n");
            printf("Energy consumed: %g Joules \n", power_printEnergy(&pDataDram));
            printf("Power consumed: %g Watts \n", power_printEnergy(&pDataDram) / runtime );
        }
    }

    if ( cpuid_hasFeature(TM2) )
    {
        thermal_init(0);
        printf("Current core temperatures:\n");

        for (uint32_t i = 0; i < cpuid_topology.numCoresPerSocket; i++ )
        {
            printf("Core %d: %u C\n",
                    numa_info.nodes[socketId].processors[i],
                    thermal_read(numa_info.nodes[socketId].processors[i]));
        }
    }

    msr_finalize();
    return EXIT_SUCCESS;
}

File: trunk/src/applications/likwid-powermeter.lua
#!<PREFIX>/bin/likwid-lua

--[[
 * =======================================================================================
 *
 *      Filename:  likwid-powermeter.lua
 *
 *      Description:  An application to get information about power 
 *      consumption on architectures implementing the RAPL interface.
 *
 *      Version:   <VERSION>
 *      Released:  <DATE>
 *
 *      Author:  Thomas Roehl (tr), thomas.roehl@gmail.com
 *      Project:  likwid
 *
 *      Copyright (C) 2014 Thomas Roehl
 *
 *      This program is free software: you can redistribute it and/or modify it under
 *      the terms of the GNU General Public License as published by the Free Software
 *      Foundation, either version 3 of the License, or (at your option) any later
 *      version.
 *
 *      This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *      WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *      PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License along with
 *      this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * =======================================================================================]]
package.path = package.path .. ';<PREFIX>/share/lua/?.lua'

local likwid = require("likwid")

local function version()
    print(string.format("likwid-powermeter --  Version %d.%d",likwid.version,likwid.release))
end

local function examples()
    print("Examples:")
    print("Measure the power consumption for 4 seconds on socket 1")
    print("likwid-powermeter -s 4 -c 1")
    print("")
    print("Use it as wrapper for an application to measure the energy for the whole execution")
    print("likwid-powermeter -c 1 ./a.out")
end

local function usage()
    version()
    print("A tool to print power and clocking information on x86 CPUs.\n")
    print("Options:")
    print("-h\t\t Help message")
    print("-v\t\t Version information")
    print("-M <0|1>\t Set how MSR registers are accessed, 0=direct, 1=msrd")
    print("-c <list>\t Specify sockets to measure")
    print("-i\t\t Print information from MSR_PKG_POWER_INFO register and Turbo mode")
    print("-s <duration>\t Set measure duration in us, ms or s. (default 2s)")
    print("-p\t\t Print dynamic clocking and CPI values")
    print("-r\t\t Print current temperatures of all CPU cores")
    print("")
    examples()
end

local config = likwid.getConfiguration();

print_info = false
use_perfctr = false
stethoscope = false
fahrenheit = false
print_temp = false
verbose = 0
if config["daemonMode"] < 0 then
    access_mode = 1
else
    access_mode = config["daemonMode"]
end
time_interval = 2.E06
use_sleep = true
sockets = {0}
eventString = "PWR_PKG_ENERGY:PWR0,PWR_PP0_ENERGY:PWR1,PWR_DRAM_ENERGY:PWR3"

cpuinfo = likwid.getCpuInfo()
cputopo = likwid.getCpuTopology()
numatopo = likwid.getNumaInfo()
affinity = likwid_getAffinityInfo()

for opt,arg in likwid.getopt(arg, "V:c:hiM:ps:vft") do
    if (opt == "h") then
        usage()
        os.exit(0)
    elseif (opt == "v") then
        version()
        os.exit(0)
    elseif (opt == "c") then
        num_sockets, sockets = likwid.sockstr_to_socklist(arg)
        if num_sockets == 0 then
            os.exit(1)
        end
    elseif (opt == "M") then
        access_mode = tonumber(arg)
        if (access_mode == nil) then
            print("Access mode (-M) must be an number")
            usage()
            os.exit(1)
        elseif (access_mode < 0) or (access_mode > 1) then
            print(string.format("Access mode (-M) %d not valid.",access_mode))
            usage()
            os.exit(1)
        end
        
    elseif (opt == "i") then
        print_info = true
    elseif (opt == "p") then
        use_perfctr = true
    elseif (opt == "f") then
        fahrenheit = true
    elseif (opt == "t") then
        print_temp = true
    elseif (opt == "V") then
        verbose = tonumber(arg)
        likwid.setVerbosity(verbose)
    elseif (opt == "s") then
        time_interval, use_sleep = likwid.parse_time(arg)
        stethoscope = true
    end
end



cpulist = {}
before = {}
after = {}
for i,socketId in pairs(sockets) do
    local affinityID = "S"..tostring(socketId)
    for j, domain in pairs(affinity["domains"]) do
        if domain["tag"] == affinityID then
            table.insert(cpulist,domain["processorList"][1])
            before[domain["processorList"][1]] = {}
            before[domain["processorList"][1]][0] = 0
            before[domain["processorList"][1]][1] = 0
            before[domain["processorList"][1]][2] = 0
            before[domain["processorList"][1]][3] = 0
            after[domain["processorList"][1]] = {}
            after[domain["processorList"][1]][0] = 0
            after[domain["processorList"][1]][1] = 0
            after[domain["processorList"][1]][2] = 0
            after[domain["processorList"][1]][3] = 0
        end
    end
end


if likwid.setAccessClientMode(access_mode) ~= 0 then
    os.exit(1)
end

power = likwid.getPowerInfo()
if not power then
    os.exit(1)
end



print(likwid.hline);
print(string.format("CPU name:\t%s", cpuinfo["name"]))
if cpuinfo["clock"] > 0 then
    print(string.format("CPU clock:\t%3.2f GHz",cpuinfo["clock"] *  1.E-09))
else
    print(string.format("CPU clock:\t%3.2f GHz",likwid.getCpuClock() *  1.E-09))
end
print(likwid.hline);

if print_info or verbose > 0 then
    if (power["turbo"]["numSteps"] > 0) then
        print(string.format("Base clock:\t%.2f MHz", power["baseFrequency"]))
        print(string.format("Minimal clock:\t%.2f MHz", power["minFrequency"]))
        print("Turbo Boost Steps:")
        for i,step in pairs(power["turbo"]["steps"]) do
            print(string.format("C%d %.2f MHz",i-1,power["turbo"]["steps"][i]))
        end
    end
    print(likwid.hline)
end

if power["hasRAPL"] == 0 then
    print("Measuring power is not supported on this machine")
    os.exit(1)
end

if (print_info) then
    print(string.format("Thermal Spec Power: %g Watts",power["tdp"]))
    print(string.format("Minimum Power: %g Watts",power["minPower"]))
    print(string.format("Maximum Power: %g Watts",power["maxPower"]))
    print(string.format("Maximum Time Window: %g micro sec",power["maxTimeWindow"]))
    print(likwid.hline)
    os.exit(0)
end

if (stethoscope) and (time_interval < power["timeUnit"]*1.E06) then
    print("Time interval too short, minimum measurement time is "..tostring(power["timeUnit"]*1.E06).. " us")
    os.exit(1)
end

if (use_perfctr) then
    affinity = likwid.getAffinityInfo()
    argString = ""
    for i,socket in pairs(sockets) do
        argString = argString .. string.format("S%u:0-%u",socket,(cputopo["numCoresPerSocket"]*cputopo["numThreadsPerCore"])-1)
        if (i < #sockets) then
            argString = argString .. "@"
        end
    end
    likwid.init(likwid.tablelength(cpulist), cpulist)
    group = likwid.addEventSet(eventString)
    likwid.setupCounters(group)
end

local execString = ""
if #arg == 0 then
    stethoscope = true
else
    for i=1,#arg do
        execString = execString .. arg[i] .. " "
    end
end

if (use_perfctr) then
    likwid.startCounters()
else
    for i,socket in pairs(sockets) do
        cpu = cpulist[i]
        if (power["hasRAPL_PP0"]) then before[cpu][1] = likwid.startPower(cpu, 1) end
        if (power["hasRAPL_PP1"]) then before[cpu][2] = likwid.startPower(cpu, 2) end
        if (power["hasRAPL_DRAM"]) then before[cpu][3] = likwid.startPower(cpu, 3) end
        before[cpu][0] = likwid.startPower(cpu, 0)
    end
end
time_before = likwid.startClock()
if (stethoscope) then
    if (use_sleep) then
        sleep(time_interval/1.E06)
    else
        usleep(time_interval)
    end
else
    err = os.execute(execString)
    if (err == false) then
        print(string.format("Failed to execute %s!",execString))
        likwid.stopCounters()
        likwid.finalize()
        likwid.putNumaInfo()
        likwid.putAffinityInfo()
        likwid.putTopology()
        os.exit(1)
    end
end
time_after = likwid.stopClock()

if (use_perfctr) then
    likwid.stopCounters()
    likwid.finalize()
else
    for i,socket in pairs(sockets) do
        cpu = cpulist[i]
        after[cpu][0] = likwid.stopPower(cpu, 0)
        if (power["hasRAPL_PP0"]) then after[cpu][1] = likwid.stopPower(cpu, 1) end
        if (power["hasRAPL_PP1"]) then after[cpu][2] = likwid.stopPower(cpu, 2) end
        if (power["hasRAPL_DRAM"]) then after[cpu][3] = likwid.stopPower(cpu, 3) end
    end
end
runtime = likwid.getClock(time_before, time_after)
print(string.format("Runtime: %g s",runtime))

for i,socket in pairs(sockets) do
    cpu = cpulist[i]
    print(string.format("Measure for socket %d on cpu %d", socket,cpu ))
    print("Domain: PKG")
    energy = likwid.calcPower(before[cpu][0], after[cpu][0], 0)
    print(string.format("Energy consumed: %g Joules",energy))
    print(string.format("Power consumed: %g Watts",energy/runtime))
    if (power["hasRAPL_PP0"]) then
        print("Domain: PP0")
        energy = likwid.calcPower(before[cpu][1], after[cpu][1], 1)
        print(string.format("Energy consumed: %g Joules",energy));
        print(string.format("Power consumed: %g Watts",energy/runtime))
    end
    if (power["hasRAPL_PP1"]) then
        print("Domain: PP1")
        energy = likwid.calcPower(before[cpu][2], after[cpu][2], 2)
        print(string.format("Energy consumed: %g Joules",energy));
        print(string.format("Power consumed: %g Watts",energy/runtime))
    end
    if (power["hasRAPL_DRAM"]) then
        print("Domain: DRAM")
        energy = likwid.calcPower(before[cpu][3], after[cpu][3], 3)
        print(string.format("Energy consumed: %g Joules",energy));
        print(string.format("Power consumed: %g Watts",energy/runtime))
    end
end
if print_temp and (string.find(cpuinfo["features"],"TM2") ~= nil) then
    print(likwid.hline)
    likwid.initTemp(cpulist[i]);
    print("Current core temperatures:");
    for i=1,cputopo["numSockets"] do
        local tag = "S" .. tostring(i-1)
        for _, domain in pairs(affinity["domains"]) do
            if domain["tag"] == tag then
                for j=1,#domain["processorList"] do
                    local cpuid = domain["processorList"][j]
                    if (fahrenheit) then
                        print(string.format("Socket %d Core %d: %u F",i-1,cpuid, 1.8*likwid.readTemp(cpuid)+32));
                    else
                        print(string.format("Socket %d Core %d: %u C",i-1,cpuid, likwid.readTemp(cpuid)));
                    end
                end
            end
        end
    end
end
    


--likwid.finalize();
likwid.putNumaInfo()
likwid.putAffinityInfo()
likwid.putTopology()
likwid.putConfiguration()
